app/lib/shared/utils.dart

```diff
@@ -361,3 +362,40 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = <Future>[];
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        while (futures.length > concurrency) {
+          await Future.any(futures.map(
+            (e) => e.whenComplete(() => futures.remove(e)),
+          ));
```

Comment:
You end up doing this `.whenComplete` a LOT of times.

Since you'll do this `O(n^2)` times, is that really necessary?

We'll remove each future from the list `O(n^2)` times. Given that list is `n` long and the time to remove an item from the list is probably `O(n)`, I'm guessing this will be something like `O(n^3)`.